app/routes.py

```diff
@@ -1,2 +1,179 @@
+def handle_tasks():
+    if request.method == "POST":
+        request_body = request.get_json()
+        if (
+            "title" not in request_body or 
+            "description" not in request_body or 
+            "completed_at" not in request_body
+        ):
+            return jsonify({"details": "Invalid data"}), 400
+
+        new_task = Task(
+            title=request_body["title"],
+            description=request_body["description"],
+            completed_at=request_body["completed_at"]
+        )
+
+        db.session.add(new_task)
+        db.session.commit()
+
+        return jsonify({"task": new_task.task_dict()}), 201
+
+    elif request.method == "GET":
+        sort_query = request.args.get("sort")
+        if sort_query == "asc":
+            tasks = Task.query.order_by(asc("title"))
+        elif sort_query == "desc":
+            tasks = Task.query.order_by(desc("title"))
+        else:
+            tasks = Task.query.all()
+
+        task_response = [task.task_dict() for task in tasks]
+
+        return jsonify(task_response), 200
+
+
+@tasks_bp.route("/<task_id>", methods={"GET", "PUT", "DELETE"})
+def handle_one_task(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(task), 404
+
+    if request.method == "GET":
+        if task.goal_id:
+            return jsonify({"task": task.task_and_goal_dict()})
+        return ({"task": task.task_dict()}), 200
+
+    elif request.method == "PUT":
+        request_body = request.get_json()
+        task.title=request_body["title"]
+        task.description=request_body["description"]
+
+        db.session.commit()
+        return jsonify({"task": task.task_dict()}), 200
+
+    elif request.method == "DELETE":
+        db.session.delete(task)
+        db.session.commit()
+        response_body = f'Task {task.task_id} "{task.title}" successfully deleted'
+        return jsonify({"details": response_body}), 200
+
+
+@tasks_bp.route("<task_id>/mark_complete", methods=["PATCH"])
+def mark_task_complete(task_id):
+    task= Task.query.get(task_id)
+    if task is None:
+        return jsonify(task), 404
+
+    current_time = datetime.now()
+    task.completed_at = current_time
+
+    BOT_TOKEN = os.environ.get('BOT_TOKEN')
+    requests.post(
+        url="https://slack.com/api/chat.postMessage", 
+        headers={"Authorization": f"Bearer {BOT_TOKEN}"}, 
+        data={"channel": "task_notifications", "text": f"Someone just completed the task {task.title}"}
+    )
+
+    db.session.commit()
+
+    return jsonify({"task": task.task_dict()}), 200
+
+
+@tasks_bp.route("<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete(task_id):
+    task = Task.query.get(task_id)
+    if task is None:
+        return jsonify(task), 404
+    task.completed_at = None
+
+    db.session.commit()
+
+    return jsonify({"task": task.task_dict()}), 200
+
+
+@goals_bp.route("", methods=["GET", "POST"])
+def handle_goals():
```

Comment:
Similar feedback about splitting these functions, and moving validation and dictionary-handling logic around that I made for `Task` will also apply to `Goal`. These are common themes for any model with CRUD operations.